File: python/90_子集II_medium.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        nums.sort()
        for mask in range(0, 1<<n):
            t = []
            flag = 1
            for i in range(n):
                if mask & (1<<i):
                    if i>0 and (mask>>(i-1) & 1)==0 and nums[i] == nums[i-1]:
                        flag = 0
                        break
                    t.append(nums[i])
            if flag:
                ans.append(t)
        return ans

File: cpp/90_子集II_medium.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> t;
        vector<vector<int>> ans;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for (int mask=0; mask<(1<<n); mask++) {
            t.clear();
            int flag = 1;
            for (int i=0; i<n; i++) {
                if (mask & (1<<i)) {
                    if (i>0 && ((mask>>(i-1)) & 1) == 0 && nums[i] == nums[i-1]) {
                        flag = 0;
                        break;
                    }
                    t.push_back(nums[i]);
                }
            }
            if (flag)
                ans.push_back(t);
        }
        return ans;
    }
};